File: backend/schemas/behavior.py
from datetime import datetime

from pydantic import BaseModel, Field


class BehaviorEventCreate(BaseModel):
    user_id: int
    action: str = Field(min_length=2, max_length=40)
    query: str | None = None
    product_id: int | None = None
    score: float = Field(default=1.0, ge=0)
    session_id: str | None = None
    context_json: str | None = None


class BehaviorEventOut(BaseModel):
    id: int
    user_id: int
    action: str
    query: str | None
    product_id: int | None
    score: float
    session_id: str | None
    context_json: str | None
    created_at: datetime

    class Config:
        from_attributes = True

File: src/services/behavior.ts
import { API_BASE_URL } from "../config/api";

type BehaviorAction = "search" | "view" | "click" | "wishlist" | "cart" | "purchase" | "feedback";

type TrackBehaviorInput = {
  userId?: number;
  action: BehaviorAction;
  productId?: number;
  query?: string;
  score?: number;
  sessionId?: string;
  context?: Record<string, unknown>;
};

const getStoredUserId = (): number => {
  if (typeof window === "undefined") {
    return 1;
  }

  const raw = localStorage.getItem("user");
  if (!raw) {
    return 1;
  }

  try {
    const parsed = JSON.parse(raw) as { id?: string };
    const numericId = Number.parseInt(parsed.id ?? "", 10);
    return Number.isFinite(numericId) ? numericId : 1;
  } catch {
    return 1;
  }
};

const getSessionId = (): string => {
  if (typeof window === "undefined") {
    return "server-session";
  }

  const existing = localStorage.getItem("behaviorSessionId");
  if (existing) {
    return existing;
  }

  const generated = `sess-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
  localStorage.setItem("behaviorSessionId", generated);
  return generated;
};

export const trackBehaviorEvent = async (input: TrackBehaviorInput): Promise<void> => {
  try {
    await fetch(`${API_BASE_URL}/behavior/track`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        user_id: input.userId ?? getStoredUserId(),
        action: input.action,
        query: input.query,
        product_id: input.productId,
        score: input.score ?? 1,
        session_id: input.sessionId ?? getSessionId(),
        context_json: input.context ? JSON.stringify(input.context) : null,
      }),
    });
  } catch {
    // Tracking should never block UX.
  }
};
